### PR TITLE
prov/psm: add PSM2 support

### PR DESCRIPTION
**Pull request by j-xiong**
(1) use a common code bases for both PSM and PSM2;
(2) different provider names are used: "psm" and "psm2";
(3) the psm and psm2 providers can be enabled in the same build,
    as long as the proper path is specified in the "--enable"
    options;
(4) at run time, only one of the two providers can succeed in
    initialization, depending on the actual PSM library in use.

Sigoed-off-by: jianxin xiong <jianxin.xiong@intel.com>

### CODE REVIEW

File: prov/psm2/src/makefile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+setup:
+	cp ../../psm/src/psmx*.[ch] .
```
Comment (jsquyres) [inline prov/psm2/src/makefile]: This seems like a bad idea.  Can you make sym links, instead?  Sym links can be committed to git. You can probably get rid of this "makefile" then...?

File: man/fi_psm.7.md [hunk 2]
```diff
@@ -21,6 +21,18 @@ tag-matching message queue functions and the Active Message functions
 to support a variety of libfabric data transfer APIs, including tagged
 message queue, message queue, RMA, and atomic operations.
 
+The *psm2* provider runs over the PSM 2.0 interface that is
```
Comment (jsquyres) [inline man/fi_psm.7.md]: You might want to add a `man/man7/fi_psm2.7` file that just contains

```man
.so man7/fi_psm.7
```

That way, people can do `man fi_psm` or `man fi_psm2`, and they'll see the same man page.

File: prov/psm2/src/makefile (hunk 1 above)
Comment (j-xiong) [inline prov/psm2/src/makefile]: @jsquyres The next line makes symbol renaming on the files. Symbolic links wouldn't work.

File: man/fi_psm.7.md (hunk 2 above)
Comment (j-xiong) [inline man/fi_psm.7.md]: Good idea, will do.

File: prov/psm2/src/makefile (hunk 1 above)
Comment (jsquyres) [inline prov/psm2/src/makefile]: Ah, I missed that.

How about having a `psm-rename.h` file that looks something like:

```c
#ifndef PSM_RENAME_H
#define PSM_RENAME_H

#ifdef PSMX_CONVERT_TO_PSMX2
#define psmx_foo psmx2_foo
#define psmx_bar psmx2_bar
// ...etc.
#endif

#endif
```

We do similar things in the OMPI code base for cases analogous to this.

Comment (j-xiong) [inline prov/psm2/src/makefile]: That sounds a reasonable approach. I will try that out.

**Comment by j-xiong** [comment]
I have updated the PR with new ways of sharing code between psm and psm2 provider. @jsquyres How do think about the change?

File: Makefile.am [hunk 3]
```diff
@@ -64,6 +66,7 @@ endif
 src_libfabric_la_CPPFLAGS = $(AM_CPPFLAGS)
 src_libfabric_la_LDFLAGS =
 src_libfabric_la_LIBADD =
+src_libfabric_la_DEPENDENCIES = $(srcdir)/libfabric.map
```
Comment (jsquyres) [inline Makefile.am]: Should this be part of a different commit or PR?  Doesn't seem related to PSM/PSM2.

File: Makefile.am [hunk 4]
```diff
@@ -294,18 +297,63 @@ libpsmx_fi_la_LDFLAGS = \
 libpsmx_fi_la_LIBADD = $(linkback) $(psm_LIBS)
 libpsmx_fi_la_DEPENDENCIES = $(linkback)
 else !HAVE_PSM_DL
-src_libfabric_la_SOURCES += $(_psm_files)
-src_libfabric_la_CPPFLAGS += $(psm_CPPFLAGS)
-src_libfabric_la_LDFLAGS += $(psm_LDFLAGS)
-src_libfabric_la_LIBADD += $(psm_LIBS)
+noinst_LTLIBRARIES += libpsmx.la
```
Comment (jsquyres) [inline Makefile.am]: Just curious: is there a reason to make this (and `libpsmx2.la`) separate convenience LT libraries (vs. just including the sources directly in the `src_libfabric_la_SOURCES`)?

**Comment by jsquyres** [comment]
@j-xiong I think the sym links + `rename.h` + per-provider `version.h` + the `update-rename-h.sh` script worked out very nicely!

File: Makefile.am (hunk 3 above)
Comment (j-xiong) [inline Makefile.am]: That line is moved from back of the file to this place because the PSM2 changes need to add items to this variable.

File: Makefile.am (hunk 4 above)
Comment (j-xiong) [inline Makefile.am]: Yes. The use of convenience libraries allows setting different CPPFLAGS and thus using different psm headers for PSM1 and PSM2.

File: Makefile.am (hunk 3 above)
Comment (jsquyres) [inline Makefile.am]: Ah, I see it now.  Got it.

File: Makefile.am (hunk 4 above)
Comment (jsquyres) [inline Makefile.am]: Thanks for the explanation.  Works for me.
